### ISSUE TITLE
Support for .NET 9

### ISSUE THREAD

**Issue by stevejgordon** [open-telemetry/opentelemetry-dotnet-instrumentation#3766]
# Feature Request

**Is your feature request related to a problem? Please describe.**

Enabling auto instrumentation for a .NET 9 application currently fails as there is no `net9.0` directory in the store.

```
[StartupHook] [2024-11-13T09:19:46.9432178Z] [Warning] Rule Engine: Couldn't evaluate reference to runtime store assemblies in an app.
Exception: Could not find a part of the path '/app/otel/store/x64/net9.0'.
System.IO.DirectoryNotFoundException: Could not find a part of the path '/app/otel/store/x64/net9.0'.
   at System.IO.Enumeration.FileSystemEnumerator`1.CreateDirectoryHandle(String path, Boolean ignoreNotFound)
   at System.IO.Enumeration.FileSystemEnumerator`1.Init()
   at System.IO.Enumeration.FileSystemEnumerable`1..ctor(String directory, FindTransform transform, EnumerationOptions options, Boolean isNormalized)
   at System.IO.Enumeration.FileSystemEnumerableFactory.UserFiles(String directory, String expression, EnumerationOptions options)
   at System.IO.Directory.InternalEnumeratePaths(String path, String searchPattern, SearchTarget searchTarget, EnumerationOptions options)
   at System.IO.Directory.GetFiles(String path, String searchPattern, EnumerationOptions enumerationOptions)
   at OpenTelemetry.AutoInstrumentation.RulesEngine.RuntimeStoreDiagnosticRule.Evaluate() in /_/src/OpenTelemetry.AutoInstrumentation.StartupHook/RulesEngine/RuntimeStoreDiagnosticRule.cs:line 41
```

**Describe the solution you'd like**

Support for .NET 9

**Additional context**

Is there a timeframe for supporting .NET 9 now that it has gone GA?

**Comment by Kielek** [comment]
@stevejgordon, first we need to do some work on https://github.com/open-telemetry/opentelemetry-dotnet-contrib repository. I hope, most of this things will be ready next week.

The tentative schedule for the release this project is end of November.

Adding this to Milestone 1.10.0.

**Comment by stevejgordon** [comment]
Okay, thanks, @Kielek. If there's any low-hanging fruit I can assist with next week, I'd likely have time to contribute. 

I need to investigate further, but I ran into issues with auto instrumentation after updating our WIP distro to 1.10, which includes a plugin we register for the auto instrumentation. I've not looked deep enough yet, but my theory is that the bundling of 1.9 in the profiler might be causing the mismatch and leading to a TypeLoadException for our code. I'll wait for the 1.10 and try again.

**Comment by pellared** [comment]
Possibly related issue https://github.com/open-telemetry/opentelemetry-dotnet/issues/5973

**Comment by Kielek** [comment]
@pellared, the root cause of the issue, is the lack of the _additional store_ in /app/otel/store/x64/net9.0. In this folder you should have already net6.0/net7.0/net8.0.

@stevejgordon, there is couple things we should change, most of them are related to our build-pipeline/dependencies. Please check following list:

* Change the linux native code builder from Centos to Ubuntu (the exact version should be checked with the lowest supported version by .NET8.0)
* Remove Centos from the build pipeline
* Replace .net6 and .net7 by .net9 in pipeline/target frameworks
* Update OTel .NET SDK to 1.10.0.
* Take new dependencies from OTel-contrib repository, all of them should be compiled against OTel .NET SDK 1.10.0 (it will ensure that works correctly with .NET9).

Nice to have maintenance tasks in OTel .NET Contrib: 
* https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/2271 (we can make a releases without this, but it will be great to cleanup the code).

Blockers:
* https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/2313
* https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/2314
* https://github.com/open-telemetry/opentelemetry-dotnet-contrib/issues/2071

Probably some other stuff.

I think that it for you, it will be easier to help on contrib side. Like https://github.com/open-telemetry/opentelemetry-dotnet-contrib/issues/2071.

I will be happy to include you in all PRs in this repository also. We will need a help in the next year for .NET10 support. It will be easier to follow examples (next week probably).

Please let me know what do you think about this.

**Comment by stevejgordon** [comment]
That sounds good, @Kielek. I've commented on https://github.com/open-telemetry/opentelemetry-dotnet-contrib/issues/2071 as someone had grabbed it but I don't see any progress since. I'll get familiar with this repo and I'm happy to be included on any PRs etc.

**Comment by stevejgordon** [comment]
@Kielek For those other items, where there are no issues. Are you happy for me to create issues to be added to a 1.10.0 milestone and tackle some of the hopefully simpler ones!

**Comment by Kielek** [comment]
@stevejgordon, sorry, I have missed you last comment.
Some update in this topic, I have problematic cases #3839 and #3838. If you have time to investigate the root cause/create PR with the fixes. It is more than welcomed.

**Comment by Kielek** [comment]
SIG meeting note: hopefully next week - make a `beta.1` release to make it available for the end-users.
Final version expected early next month (year). We would like to avoid any problematic stuff during the holidays.

**Comment by Kielek** [comment]
@stevejgordon, released as a beta [1.10.0-beta.1](https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/pull/3887).

Please let me know if it is working for your, do you have any feedback, etc.

**Comment by stevejgordon** [comment]
Thanks so much, @Kielek. I'll take it for a spin the first week after the holidays.
